File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []

        digit_to_char = {
            '2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
            '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'
        }
        
        result = []
        
        def backtrack(index: int, current_combination: str):
            if index == len(digits):
                result.append(current_combination)
                return
            
            current_digit = digits[index]
            for letter in digit_to_char[current_digit]:
                backtrack(index + 1, current_combination + letter)
        
        backtrack(0, "")
        return result

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty()) return {};

        vector<string> digitToChar = {
            "",    // 0 (not used)
            "",    // 1 (not used)
            "abc", // 2
            "def", // 3
            "ghi", // 4
            "jkl", // 5
            "mno", // 6
            "pqrs",// 7
            "tuv", // 8
            "wxyz" // 9
        };

        vector<string> result; 
        string currentCombination; 

        
        function<void(int)> backtrack = [&](int index) {
            if (index == digits.size()) {
                result.push_back(currentCombination);
                return;
            }

            string letters = digitToChar[digits[index] - '0'];
            for (char letter : letters) {
                currentCombination.push_back(letter); 
                backtrack(index + 1);
                currentCombination.pop_back(); 
            }
        };

        backtrack(0); 
        return result;
    }
};